### PR TITLE
Adds `autoIncrease` flag to increase authorizations in `topUp`

### PR DESCRIPTION
**Pull request by vzotova**
Based on #153 
With flag set to true each `topUp` will increase authorization amount for all authorized app

### CODE REVIEW

**Review by cygnusv — APPROVED** [review]
LGTM! Just one comment: Why using the "toggle" approach instead of a setter? It's slightly less ergonomic, in my opinion, but interested to see what was your line of thinking.

File: contracts/staking/IStaking.sol [hunk 1]
```diff
@@ -177,19 +158,15 @@ interface IStaking {
...
-    ///         the owner or the staking provider.
-    function topUpKeep(address stakingProvider) external;
-
-    /// @notice Propagates information about stake top-up from the legacy NU
-    ///         staking contract to T staking contract. Can be called only by
-    ///         the owner or the staking provider.
-    function topUpNu(address stakingProvider) external;
+    /// @notice Toggle auto authorization increase flag. If true then all amount
+    ///         in top-up will be added to already authorized applications.
+    function toggleAutoAuthorizationIncrease(address stakingProvider) external;
```
[the remark this answers is not in the document]
Comment (vzotova) [reply contracts/staking/IStaking.sol:169]: I'm not sure about that, I want names explicit as possible

**Comment by vzotova** [comment]
> LGTM! Just one comment: Why using the "toggle" approach instead of a setter? It's slightly less ergonomic, in my opinion, but interested to see what was your line of thinking.

not strong opinion: from code perspective - if you call this tx you want to change parameter/toggle, there is no other options that you can change except to opposite value. In the same case setter should fail when parameter is equal.

**Review by cygnusv — APPROVED** [review]

File: contracts/staking/TokenStaking.sol [hunk 2]
```diff
@@ -590,6 +597,54 @@ contract TokenStaking is Initializable, IStaking, Checkpoints {
         emit ToppedUp(stakingProvider, amount);
         increaseStakeCheckpoint(stakingProvider, amount);
         token.safeTransferFrom(msg.sender, address(this), amount);
+
+        if (!stakingProviderStruct.autoIncrease) {
+            return;
+        }
+
+        // increase authorization for all authorized app
+        for (
```
Comment (lukasz-zimnoch) [inline contracts/staking/TokenStaking.sol:606]: The `topUp` function can be called by anyone so, if `autoIncrease` is `true`, anyone can increase staking provider's authorization. Of course, the caller has to transfer an amount, but still, wondering if we aren't opening a potential attack vector here. Shouldn't we do the increase just for callers being provider's authorizers/owners? Curious about your thoughts!

Comment (cygnusv) [reply contracts/staking/TokenStaking.sol:606]: Yes, this was intentionally added in https://github.com/threshold-network/solidity-contracts/pull/110.
The rationale was to allow other contracts like Merkle distribution or application contracts to automatically top up stakes.

Comment (lukasz-zimnoch) [reply contracts/staking/TokenStaking.sol:606]: Ok, thanks for clarification!

**Review by lukasz-zimnoch — DISMISSED** [review]

**Review by theref — APPROVED** [review]

**Review by cygnusv — APPROVED** [review]

**Pull request merged**